transports/tcp/src/provider/tokio.rs

```diff
@@ -0,0 +1,160 @@
+        -> Poll<io::Result<Incoming<Self::Stream>>>
+    {
+        let (stream, remote_addr) = match l.poll_accept(cx) {
+            Poll::Pending => return Poll::Pending,
+            Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+            Poll::Ready(Ok((stream, remote_addr))) => (stream, remote_addr)
+        };
+
+        let local_addr = stream.local_addr()?;
+        let stream = TcpStream(stream);
+
+        Poll::Ready(Ok(Incoming { stream, local_addr, remote_addr }))
+    }
+
+    fn poll_interfaces(w: &mut Self::IfWatcher, cx: &mut Context<'_>) -> Poll<io::Result<IfEvent>> {
+        loop {
+            if let Some(event) = w.pending.pop() {
+                return Poll::Ready(Ok(event))
+            }
+
+            match Pin::new(&mut w.delay).poll(cx) {
+                Poll::Pending => return Poll::Pending,
+                Poll::Ready(()) => {
+                    let ifs = get_if_addrs()?;
+                    let addrs = ifs.into_iter().map(|iface| match iface.addr {
+                        IfAddr::V4(ip4) => {
+                            let prefix_len = (!u32::from_be_bytes(ip4.netmask.octets())).leading_zeros();
+                            let ipnet = Ipv4Net::new(ip4.ip, prefix_len as u8)
+                                .expect("prefix_len can not exceed 32");
+                            IpNet::V4(ipnet)
+                        }
+                        IfAddr::V6(ip6) => {
+                            let prefix_len = (!u128::from_be_bytes(ip6.netmask.octets())).leading_zeros();
+                            let ipnet = Ipv6Net::new(ip6.ip, prefix_len as u8)
+                                .expect("prefix_len can not exceed 128");
+                            IpNet::V6(ipnet)
+                        }
+                    }).collect::<HashSet<_>>();
+
+                    for down in w.addrs.difference(&addrs) {
+                        w.pending.push(IfEvent::Down(*down));
+                    }
+
+                    for up in addrs.difference(&w.addrs) {
+                        w.pending.push(IfEvent::Up(*up));
+                    }
+
+                    w.addrs = addrs;
+                    w.delay.reset(Duration::from_secs(10));
+                }
+            }
+        }
+    }
+}
+
+/// A [`tokio_crate::net::TcpStream`] that implements [`AsyncRead`] and [`AsyncWrite`].
+#[derive(Debug)]
+pub struct TcpStream(pub tokio_crate::net::TcpStream);
+
+impl Into<tokio_crate::net::TcpStream> for TcpStream {
+    fn into(self: TcpStream) -> tokio_crate::net::TcpStream {
+        self.0
+    }
+}
+
+impl AsyncRead for TcpStream {
+    fn poll_read(mut self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<Result<usize, io::Error>> {
+        let mut read_buf = tokio_crate::io::ReadBuf::new(buf);
+        futures::ready!(tokio_crate::io::AsyncRead::poll_read(Pin::new(&mut self.0), cx, &mut read_buf))?;
+        Poll::Ready(Ok(read_buf.filled().len()))
+    }
+}
+
+impl AsyncWrite for TcpStream {
```

Comment:
It would be nice to forward `poll_write_vectored`, which is very helpful for rustls.

Reply:
Done in https://github.com/libp2p/rust-libp2p/pull/1887/commits/07fd1fd1c5f2c164969c163bf7fa25c42303274d.